examples/reflection/custom_attributes.rs

```diff
@@ -0,0 +1,90 @@
+//! Demonstrates how to register and access custom attributes on reflected types.
+
+use bevy::reflect::{Reflect, TypeInfo, Typed};
+use std::any::TypeId;
+use std::ops::RangeInclusive;
+
+fn main() {
+    // Bevy supports statically registering custom attribute data on reflected types,
+    // which can then be accessed at runtime via the type's `TypeInfo`.
+    // Attributes are registered using the `#[reflect(@...)]` syntax,
+    // where the `...` is any expression that resolves to a value which implements `Reflect`.
+    // Note that these attributes are stored based on their type:
+    // if two attributes have the same type, the second one will overwrite the first.
+
+    // Here is an example of registering custom attributes on a type:
+    #[derive(Reflect)]
+    struct Slider {
+        #[reflect(@RangeInclusive::<f32>::new(0.0, 1.0))]
+        // Alternatively, we could have used the `0.0..=1.0` syntax,
+        // but remember to ensure the type is the one you want!
+        #[reflect(@0.0..=1.0_f32)]
```

Comment:
Supporting this case so ergonomically (and "for free" from a complexity standpoint thanks to the Expr parse) is excellent :)

Reply:
Some might think its _too ergonomic_ as it throws out the "range" and/or min/max context. But I like that people have the choice.

Reply:
Yep! It's great that the user has some freedom in how they want to write it (as long as they type it properly).

Also, note that one downside to parsing `Expr` directly is that we can't really control what users put as the expression:

```rust
#[derive(Reflect)]
#[reflect(@{
    fn run_forever() {
        loop {}
    }

    run_forever();
    123
})]
struct Foo;
```

I think this should generally be okay. Users are unlikely to do this, and if they do, that's their decision lol. It's also as much of an attack vector as running any other code, so I don't think we need to worry about malicious use. Plus, it does allow users to get creative if they need to.

I like the idea that it can be pretty much any expression for the ergonomics factor. But if we really want to, we can always go back and limit what the user can put (this would involve a lot of manual parsing, though, if we want to have similar ergonomics).

Reply:
And it should be noted that this isn't just a custom attribute problem. You can do the same thing with serde for example:

```rust
#[derive(Serialize)]
struct Foo {
  #[serde(default = "run_forever")]
  bar: usize
}

fn run_forever() -> usize {
  loop {}
  123
}
```

Reply:
And while we could limit this by using lazily evaluation with functions or something, I was hoping it could allow users to add types with interior mutability if they so desire. Ideally these attributes are completely static, but I didn't want to be _too_ restrictive.

I can definitely change that, though. It's up to us on how much freedom we want to give users.

Reply:
Expr is actually an enum over many different expression types. We could limit it to very specific expressions if we wanted. For added "accident protection", we could also require an explicit (and derivable) ReflectAttribute trait to provide type errors for people dropping "the wrong thing" into this bucket. Although requiring a trait could also be overly limiting.

Reply:
Right, which is why I think this should be fine for an initial implementation.

Reply:
> Expr is actually an enum over many different expression types. We could limit it to very specific expressions if we wanted.

Yeah this is a possibility. Unfortunately it would require some manual parsing to check that we parse the correct expression type. I don't think you can parse select variants on their own :/